### ISSUE TITLE
HTTP: Websocket protocol handler does not allow to read independently of sending or connecting

### ISSUE THREAD

**Issue by pragmasoft-ua** [gatling/gatling#3773]
We have the following issue:

In our system (graphql subscriptions), messages are published with a regular http request, whereas subscribed with websocket, so websocket check does not need to send anything before it checks.

A case with check just after ws connection does not work either, because we need to send http request between ws connection and ws check.

**Comment by slandelle** [comment]
So so trick here is that there's a gap between the time you send your HTTP request and the time you'd start waiting for an incoming message on the WebSocket.
What if the message arrives in this gap?
We'd need to always buffer all unmatched incoming message for some time, in case a check is about to be set.
And then, how to correlate?

**Comment by pragmasoft-ua** [comment]
I think there's already some mechanism in place, otherwise it wouldn't work, when we use send empty string and then immediate check hack

Likely, at the network level, there exists some buffering of unread websocket bytes

Also match() can be used to skip those incoming messages you aren't interested in

**Comment by slandelle** [comment]
> I think there's already some mechanism in place, otherwise it wouldn't work, when we use send empty string and then immediate check hack

No, there isn't. Everything happens in the same exec and we take extra care of not missing events (which could happen in Gatling 2).

**Comment by pragmasoft-ua** [comment]
Ok, so what then happens between successive execs? If you don't buffer between successive execs, then you have problems, whether you allow reading without writing, or not. If you buffer, then you won't have problems in both these cases. 

My point is, that addition of the ability to read without writing does not change current status quo - it does not create more problems, over those which may already exist, but may help in some cases

I tried a naive way of doing something like 

```
.exec(session => {
ws.checkTextMessage("ws testSub notification received").check(jsonPath("$.payload.data.testSub").is(messageText))
session
})

```
It compiles but obviously does not work, I'm wondering if there's some way to achieve my goal without ever doing any changes in Gatling, I simply don't know scala / gatling so well. 

Maybe that check builder can be built with the session to return validation?

**Comment by slandelle** [comment]
> If you don't buffer between successive execs, then you have problems, whether you allow reading without writing, or not.

No, we don't have any problem. There are things we currently support and things we don't. What we currently support are checking against incoming messages that are triggered by a client WbeSocket action: connect or outgoing message.

> I'm wondering if there's some way to achieve my goal without ever doing any changes in Gatling

There isn't. The WebSocket support is already an awfully complex state machine. Implementing your use case would require more complexity. Doable but not a priority for us unless we get a contract for it. Or a contribution.

**Comment by pschiffe** [comment]
We would love to have this feature in Gatling as well. But I guess the only proper solution here is parallel execution - #3783 - to be able to listen to websockets in parallel to http.

**Comment by nhimanshujain** [comment]
Issue (Problem Statement): **Websocket client must receive messages from websocket server without sending any outgoing messages.** 

Sending a dummy text from client and using await to receive the message from the server does not work in my use case. This is because the server continuously sends messages to the client and using the existing approach leads to loss of some of the messages.

Code snippet to show where the loss of messages happen:
```
val keepAliveMsg = "{\"keepalive\":\"true\"}"
val awaitTime = 1.toMicros.microseconds
val checkType = ws.checkTextMessage("receive-websocket")
      .matching(jsonPath("$.data.type").is("${messageType}"))
      .check(jsonPath("$.type").is("StationLogin"))

ws("send-keepalive").sendText(keepAliveMsg).await(awaitTime)(checkType) // Loss of messages due to await
```

This feature is critical for load testing one of the services that is used by large scale users. 

I require support from the Gatling community in this regards. If the above use case is already addressed, please point me in that direction and if not, when can we expect this feature in the Gatling.
Thanks in advance!!

CC: @slandelle

**Comment by slandelle** [comment]
See #4529
